### DISCUSSION TITLE
sometime exceptions

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#10090] [Using MicroPython]
I keep having these curious little episodes with thonny & other repls where errors go unnoticed by the interpreter & it's not until I save the file to flash & let it run from there that the error is flagged. In the most recent episode on an rpi pico I had 
```
v=machine.ADC(29).read_u16()
```
in the program but hadn't imported machine at the top, where I had only
```
from machine import Pin
```
This worked in the repl, only got the exception after I'd saved the file as boot.py in flash & it tried to run from flash on power up.

Does anybody else get this behaviour & what might be the cause?

**Comment by robert-hh** [comment]
Looks like a Thonny bug/feature. To see what is caused by your code and what is caused by Thonny, try working with a dumb terminal emulator like Putty (Windows) or picocom (Linux) or screen (MAC orLinux). That skips Thonny. Like all developed IDE, Thonny is nice & convenient as long as there is no problem, and a problem amplifier as soon as something fails.

**Comment by jimmo** [comment]
Don't put anything in boot.py that isn't required for the board to boot (i.e. anything that isn't initialisation of file systems or USB configuration).

If you are using STM32 and your board is connected over USB, then USB (and hence the serial connection) won't be initialised in boot.py.

**Comment by kjm1102** [comment]
What's the thinking behind that prescription Jim? I tend to run my program from boot.py with a duplicate copy in main.py. Admittedly this is a legacy approach from the bad old days before little FS when flash memory file systems would fall over at the drop of hat, but I'm keen to know why the preference for using main.py only for user code?

**Reply by robert-hh** [reply]
> What's the thinking behind that prescription Jim

I see it more as a convention that a strict requirement, besides the state of the hardware, e.g. for USB set-up at some boards. The device can be considered as full initialized at the start of main.py, and there may be differences when boot.py starts.

**Reply by jimmo** [reply]
> What's the thinking behind that prescription Jim?

@kjm1102 Basically the reason I outlined -- boot.py configures what you need for main.py, and things like serial output or filesystem access might be be unavailable in boot.py.

If boot.py crashes then you can also end up without things (like USB) being configured.

> I tend to run my program from boot.py with a duplicate copy in main.py.

I'm not sure how this would solve anything other than making things more complicated.
